### PR TITLE
Light client friendly events

### PR DESCRIPTION
**Pull request by pepyakin**
This PR proposes a solution for light client friendly SRML events.

Each deposited event can specify zero or more topics. A topic is a fixed sized value (atm, is just a hash). When an event is deposited with some topic, the runtime adds the index of this event in the `<Events<T>>` list to a mapping that maps from topics to lists of indexes.

Light clients can "subscribe" to changes of specific storage values (see #628). The idea is that light clients would subscribe to changes of the storage values for a topic they are interested in. When the light client detects a change, it fetches the contents of the indexes list for the specific topic and with this the light client can query events' data by the indexes.

Marking this as draft, since I don't know if this is the way that we want to go, I didn't test this with light-client nor inspected the storage values and there are some questions left

### CODE REVIEW

**Comment by gavofyork** [comment]
Be good to get an opinion on this @svyatonik

**Comment by pepyakin** [comment]
I've checked the storage entries and add a test. This should be ready for review then.

File: srml/system/src/lib.rs [hunk 1]
```diff
@@ -373,6 +382,59 @@ pub fn ensure_inherent<OuterOrigin, AccountId>(o: OuterOrigin) -> Result<(), &'s
...
+		// Appending can only fail if `Events<T>` can not be decoded or
+		// when we try to insert more than `u32::max_value()` events.
+		//
+		// We perform early return if we've reached the maximum capacity of the event list,
+		// so `Events<T>` seems to be corrupted. Also, this has happened after the start of execution
+		// (since the event list is cleared at the block initialization).
+		// The most sensible thing to do here is to just ignore this event and wait until the
+		// new block.
+		if <Events<T>>::append(&[event]).is_err() {
+			// TODO(reviewers): debug_assert!(false) is tempting here, since we want to be safe in
```
Comment (pepyakin) [inline srml/system/src/lib.rs]: What do you think about this?

**Comment by pepyakin** [comment]
There might be a problem with this approach, marking it inprogress until we figure it out.

File: srml/system/src/lib.rs (hunk 1 above)
Comment (gavofyork) [reply srml/system/src/lib.rs]: i'm ok with it in, as long as it never goes anywhere near production code. the possibility of introducing a panicker in production is just too dangerous.

**Comment by gavofyork** [comment]
what might the problem be?

**Comment by pepyakin** [comment]
Yes, there is indeed a problem, although not a significant one. 

Consider the case when an event `E` was deposited with a topic `A`. 

So at the end of the block `N` the field `<EventTopics<T>>` will be like:

```
A => vec![EventIndex(0)]
```

then, let's suppose that the same event `E` with the same topic `A` was deposited at the block `N+1`. Therefore, `<EventTopics<T>>` will look the same at the end of the `N+1` block:

```
A => vec![EventIndex(0)]
```
Because the value at the storage location of `A` doesn't change there will be no notification of change between for the `N+1`. 

I am still not entirely sure, but it seems to be a problem. If the light-client doesn't follow each block, and say, it "subscribes" to changes at the block `N+1` it won't get the update.

To circumvent this problem we could store `Vec<(EventIndex, BlockNumber)>` in `EventTopics` instead of just `Vec<EventIndex>`. Or maybe even something like `Option<(BlockNumber, Vec<EventIndex>)>`.

**Comment by pepyakin** [comment]
Ok, I put up a commit which fixes this. However, I went with `Vec<(EventIndex, BlockNumber)>` instead of `Option<(BlockNumber, Vec<EventIndex>)>`. The reason for that is `::append` is not available for this type. See the commit for details:
https://github.com/paritytech/substrate/pull/2491/commits/5f9ca06de6471b48e6ea84ebf2e3cb9c48cff122

Apparently it starts asking for a hand-rolled structure tailored for this particular use-case.

**Comment by gavofyork** [comment]
ok, but please add a TODO and issue explaining that this is highly inefficient and needs to be altered to `Option<(BlockNumber, Vec<EventIndex>)>` once we can implement `::append` for it. please also assign it 1.x milestone

**Comment by xlc** [comment]
This seems like a breaking change and requires JS SDK side change as well.
@jacogr I don't think it is easy to have a JS SDK that compatible with Substrate versions that with and without this change.

**Pull request merged**